target/mips/op_helper_cheri.c

```diff
@@ -2173,28 +2173,21 @@ static void store_cap_to_memory(CPUMIPSState *env, uint32_t cs,
 {
     const cap_register_t *csp = get_readonly_capreg(&env->active_tc, cs);
     uint64_t cursor = cap_get_cursor(csp);
-
     uint64_t pesbt;
     if (csp->cr_tag)
         pesbt = compress_128cap(csp);
     else
         pesbt = csp->cr_pesbt_xored_for_mem;
 
-    /* Store pesbt to memory (might trap on store) */
-    cpu_stq_data_ra(env, vaddr, pesbt, retpc);
-    /*
-     * We can set the tag bit now because we know we won't get a tlb fault
-     * after the first store (capabilities must be on the same page).
-     */
-    /* Set the tag bit in memory, if set in the register. */
     env->statcounters_cap_write++;
```

Comment:
We should have comments in each of these methods documenting that the tag store must come first.

Reply:
I agree, we should have the comment state that cheri_tag_set/cheri_tag_invalidate will take the TLB trap first so the cpu_stq will succeed.